src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs

```diff
@@ -458,10 +564,188 @@ private HqlTreeNode TranslateEqualityComparison(BinaryExpression expression)
 					_hqlTreeBuilder.IsNull(rhs2)));
 		}
 
-		static bool IsNullable(HqlExpression original)
+		private bool IsNullable(Expression expression, BinaryExpression equalityExpression)
```

Comment:
Would it be a clearer code if it put into a expression tree visitor?

Reply:
Not really as the problem with the expression visitor is that its metohds are all void and converting for example `IsNullable(binaryExpression.Left) || IsNullable(binaryExpression.Right)` would require more code. Nonetheless, I simplified the code by removing the unneeded `while` statement, split the larger cases into separated methods and moved all logic into a separated class as `HqlGeneratorExpressionVisitor` is quite large without the nullable check.